cadc-conesearch/src/main/java/org/opencadc/conesearch/TAPQueryGenerator.java

```diff
@@ -0,0 +1,236 @@
+ *  OpenCADC project.                    OpenCADC.
+ *
+ *  OpenCADC is free software:           OpenCADC est un logiciel libre ;
+ *  you can redistribute it and/or       vous pouvez le redistribuer ou le
+ *  modify it under the terms of         modifier suivant les termes de
+ *  the GNU Affero General Public        la “GNU Affero General Public
+ *  License as published by the          License” telle que publiée
+ *  Free Software Foundation,            par la Free Software Foundation
+ *  either version 3 of the              : soit la version 3 de cette
+ *  License, or (at your option)         licence, soit (à votre gré)
+ *  any later version.                   toute version ultérieure.
+ *
+ *  OpenCADC is distributed in the       OpenCADC est distribué
+ *  hope that it will be useful,         dans l’espoir qu’il vous
+ *  but WITHOUT ANY WARRANTY;            sera utile, mais SANS AUCUNE
+ *  without even the implied             GARANTIE : sans même la garantie
+ *  warranty of MERCHANTABILITY          implicite de COMMERCIALISABILITÉ
+ *  or FITNESS FOR A PARTICULAR          ni d’ADÉQUATION À UN OBJECTIF
+ *  PURPOSE.  See the GNU Affero         PARTICULIER. Consultez la Licence
+ *  General Public License for           Générale Publique GNU Affero
+ *  more details.                        pour plus de détails.
+ *
+ *  You should have received             Vous devriez avoir reçu une
+ *  a copy of the GNU Affero             copie de la Licence Générale
+ *  General Public License along         Publique GNU Affero avec
+ *  with OpenCADC.  If not, see          OpenCADC ; si ce n’est
+ *  <http://www.gnu.org/licenses/>.      pas le cas, consultez :
+ *                                       <http://www.gnu.org/licenses/>.
+ *
+ *
+ ************************************************************************
+ */
+
+package org.opencadc.conesearch;
+
+import ca.nrc.cadc.dali.Circle;
+import ca.nrc.cadc.dali.CommonParamValidator;
+import ca.nrc.cadc.dali.tables.votable.VOTableWriter;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+
+/**
+ * Class to produce an ADQL query from a set of Parameters
+ */
+public class TAPQueryGenerator extends CommonParamValidator {
+    private static final int DEF_MAXREC = 1000;
+    private static final int MAX_MAXREC = Integer.MAX_VALUE;
+    private static final int MIN_VERB_VALUE = 1;
+    private static final int MID_VERB_VALUE = 2;
+    private static final int MAX_VERB_VALUE = 3;
+    private static final Logger LOGGER = LogManager.getLogger(TAPQueryGenerator.class);
+
+    protected final String catalog;
+    protected final String lowVerbositySelectList;
+    protected final String midVerbositySelectList;
+    protected final String highVerbositySelectList;
+
+    public TAPQueryGenerator(final String catalog, final String lowVerbositySelectList,
```

Comment:
The position column also must be a ctor arg since it is tightly coupled with the table name. `catalog` is pretty generic so I would use `tableName` and `posColumnName`.

The verb-dependent select lists I see as a pretty fragile way to pick columns, but I can't think of anything bulletproof so I guess simplest is the right choice now. I can imagine too many variations and they would all be complex :-)

This generator is a different style from the sia2/AdqlQueryGenerator; that class takes the input param map in the ctor and returns the TAP params from a no-arg `getParameterMap(). I can see you looked at that because both of them have the same incorrect comment mentioning the REQUEST param (which no longer exists in TAP). The sia2 style makes a single use generator, while this allows calling code to generate different queries with different inputs; there are no use cases for the latter since a cone search request is inherently single query.

Finally, the ctor should check all args for null and throw IllegalArgumentException if necessary.

Reply:
This is done.